[Draft] Qualcomm AI Engine Direct - Support kv_cached llama2 model

shewu-quic opened the pull request:
### Summary
- Support static kv_cached llama2 model
  - We reference AIMET jupyter notebooks and implement a [static LLAMA](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/model/static_llama.py)
- Add [qnn_llama_runner](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/executor_runner/qnn_llama_runner.cpp) to run static LLAMA
- Add [e2e example script](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/llama.py) verified with story110M

### Notes
- In fp16 mode, the model can be compiled and executed on the device to obtain accurate results. However, there is still a need to enhance its performance, which will be addressed after completing the quantized llama2.
- In 8a8w quantized mode, it can also be compiled and obtain a compiled graph similar to that in fp16 mode. However, when executed on the device, the results are not as expected.
- For now, we are going to 16 bit quantization.
- The main difference between [static LLAMA](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/model/static_llama.py) and [existent examples/models/llama2](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/models/llama2/model.py) is that we regard [kv cache as the i/o of graph](https://github.com/CodeLinaro/executorch/blob/9681c0195a17952a95042c407d4d01e4e2b4bab1/examples/qualcomm/llama2/model/static_llama.py#L50).

### Compiled graph
For now, we will fallback the following ops which are about reading and updating attention mask:
- aten_index_tensor
- aten_index_put_default
![image](https://github.com/pytorch/executorch/assets/138087975/288ad43d-7bd8-43ae-85d7-f10a228164f5)

### Prepare model
#### Download and export stories110M model
```
# tokenizer.model & stories110M.pt:
wget "https://huggingface.co/karpathy/tinyllamas/resolve/main/stories110M.pt"
wget "https://raw.githubusercontent.com/karpathy/llama2.c/master/tokenizer.model"

# tokenizer.bin:
python -m examples.models.llama2.tokenizer.tokenizer -t tokenizer.model -o tokenizer.bin

# params.json:
echo '{"dim": 768, "multiple_of": 32, "n_heads": 12, "n_layers": 12, "norm_eps": 1e-05, "vocab_size": 32000}' > params.json
```
### Run e2e example script
```
# fp16:
python examples/qualcomm/llama2/llama.py -a xxx -b build_android -s xxx -m SM8650 -F --checkpoint stories110M --params params.json --tokenizer_bin tokenizer.bin --prompt Once
# quant:
python examples/qualcomm/llama2/llama.py -a xxx -b build_android -s xxx -m SM8650 --ptq 8a8w --tokenizer_model tokenizer.model --checkpoint stories110M --params params.json --tokenizer_bin tokenizer.bin --prompt Once
```

haowhsu-quic committed:
Qualcomm AI Engine Direct - support static llama2 with kv_cache

summary
- support static kv_cached llama2 model
- add qnn_llama_runner
- add e2e example script verified with story110M

cccclai commented on backends/qualcomm/partition/common_defs.py:
```diff
@@ -11,8 +11,9 @@
     exir_ops.edge.aten.clone.default,
-    exir_ops.edge.aten.index.Tensor,
     exir_ops.edge.aten.full.default,
+    exir_ops.edge.aten.index.Tensor,
+    exir_ops.edge.aten.index_put.default,
```
Maybe these two lines can be landed seperately? It's not supported anyway and we can decouple it from this large PR

haowhsu-quic replied:
I think we still need to have it. We need `aten.index.Tensor` for slicing kv_cache and feed to individual LlamaAttention layer. `aten.index_put` is used to update kv_mask after finishing each inference.
These two operators are not supported yet by qualcomm backend, we'll need partitioner to identify them and make them fallback to CPU.

haowhsu-quic replied:
I see, we'll follow your suggestion to split PRs.

cccclai commented on backends/qualcomm/quantizer/utils.py:
```diff
@@ -119,15 +119,19 @@ def annotate_binary(node: Node, quantization_config: QuantizationConfig) -> None
     input_qspec_map = {}
     input_act0 = node.args[0]
-    if isinstance(input_act0, Node):
+    if isinstance(input_act0, Node) and input_act0.meta["val"].dtype == torch.float32:
```
We can have a seperate PR for this line too. https://github.com/pytorch/executorch/pull/2957 has a few more checks that we may need.

haowhsu-quic replied:
Yes, this looks like a common issue.

cccclai commented on backends/qualcomm/scripts/build.sh:
```diff
@@ -71,6 +71,7 @@ if [ "$BUILD_AARCH64" = true ]; then
         -DCMAKE_INSTALL_PREFIX=$BUILD_ROOT \
         -DEXECUTORCH_BUILD_QNN=ON \
         -DEXECUTORCH_BUILD_SDK=ON \
+        -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
```
Same here, a seperate PR

haowhsu-quic replied:
This is related to `qnn_llama_runner`, I think we still need to land it together.

haowhsu-quic replied:
Our custom runner ('llama2/runner/runner') invokes methods inside `extension/module/module.cpp`. We could get rid of it if required.

cccclai commented on backends/qualcomm/tests/models.py:
```diff
@@ -29,7 +29,7 @@ def __init__(self):
 
     def forward(self, x):
-        return 10.0 + x
+        return 10 + x
```
What's this for? Is it a different test case?

haowhsu-quic replied:
It's a minor fix when checking if integer type addition works. Here changes the constant to be integer type to match the function title `AddConstantLong`.

cccclai commented on examples/qualcomm/scripts/utils.py:
```diff
@@ -212,25 +228,39 @@ def build_executorch_binary(
+
+    executorch_config = ExecutorchBackendConfig(
+        extract_constant_segment=False,
+        # For shared buffer, user must pass the memory address
+        # which is allocated by RPC memory to executor runner.
+        # Therefore, won't want to pre-allocate
+        # by memory manager in runtime.
```
This seems like an optmization opportunity and we can add them later? For the alpha release, maybe let's get a functional version, and improve it step by step. What do you think?

haowhsu-quic replied:
The shared buffer has already been done by https://github.com/pytorch/executorch/pull/2531. Shared buffer mechanism is enabled for the rest of examples but not llama now (we'll have it in the near future, need changes in `qnn_llama_runner`).
For now, we still need this refactored code for other examples being functional as usual.

cccclai commented on examples/qualcomm/llama2/runner/runner.h:
```diff
@@ -0,0 +1,108 @@
+  Error generate(
+      const std::string& prompt,
+      int32_t seq_len,
+      ManagedTensor& managed_atten_mask,
+      ManagedTensor& managed_k_cache,
+      ManagedTensor& managed_v_cache,
+      ManagedTensor& managed_kv_mask,
```
It seems to me the main change is the model input, is it correct?

haowhsu-quic replied:
Yes, to make kv_cache be static shape.

cccclai commented on examples/qualcomm/llama2/model/static_llama.py:
```diff
@@ -0,0 +1,253 @@
+                v_cache=v,
+                kv_mask=kv_mask,
+            )
+            output_k_cache.append(
+                k.view(self.max_batch_size, self.max_seq_len, self.dim)
+            )
+            output_v_cache.append(
+                v.view(self.max_batch_size, self.max_seq_len, self.dim)
+            )
```
Aren't we having a dynamic shape for `output_k_cache` here?

haowhsu-quic replied:
No, we're concatenating kv_cache calculated from all attention layers:

- Input shape of k_cache: (max_batch_size, n_layers, max_seq_len, embedding_dim)
- Slice for every attention layer: (max_batch_size, max_seq_len, embedding_dim)

We could concatenate them back without runtime shape changing.

cccclai replied:
hmm probably need to check the graph, I just feel like `output_v_cache` shape is changing when iterating the layer, because we keep appending it

haowhsu-quic replied:
The list append method will be changed after `torch.export`. The output kv_caches from each attention layer are connected directly to final `concat` operator in `LlamaModel.forward`.

cccclai reviewed (commented):
Looks great! I think there are many optimization we can leverage in this pr. Just wonder if we can decouple some minor changes when getting the end to end working.

Also reminder we probably need to cherry-pick

cccclai commented:
Also curious what visualization tool you're using

digantdesai commented on examples/qualcomm/llama2/runner/runner.cpp:
```diff
@@ -0,0 +1,386 @@
+/*
```
curious if we can share this file with `models/llama2`? Just easier to maintain.

shewu-quic commented:
> Also curious what visualization tool you're using

I think we use FxGraphDrawer to visualize the graph module
https://github.com/pytorch/executorch/blob/d761f99f6fc952975835f4807a12319c121c4b90/backends/qualcomm/utils/utils.py#L136

shewu-quic commented:
> I'm trying to repro on my side. What QNN library version did you use? The error message on my side is
> 
> ```
> [ERROR] [Qnn ExecuTorch]: initial_sequencer_dp.cc:160:ERROR:A single op, "q::Concat" (Op ID: 315c00000086e), requires 0xa00000 bytes of TCM, which is greater than the TCM size of 0x800000!
> 
> [ERROR] [Qnn ExecuTorch]: initial_sequencer_dp.cc:167:ERROR:The name of the failing op before optimization is: "q::QNN_Reshape" (Op ID: 86e).
> 
> [ERROR] [Qnn ExecuTorch]: QnnDsp <E> "aten_view_copy_default_423" generated: Requires 0xa00000 bytes of TCM, which is greater than the TCM size of 0x800000!
> 
> [ERROR] [Qnn ExecuTorch]: QnnDsp <E> RouterX86 graph prepare failed 13
> 
> [ERROR] [Qnn ExecuTorch]: QnnDsp <E> Failed to finalize graph (id: 2) with err 1002
> 
> [ERROR] [Qnn ExecuTorch]: Failed to finalize Qnn Graph with error: 1002
> ```

I use QNN 2.20 and can reproduce on SM8475 from my side.

cccclai commented:
I was using qnn 2.19 and just switch to 2.20. I'm using SM8450 on my side

cccclai commented:
I was able to repro the fp version on my side, but for the 8a8w version, I hit model loading error
```
[ERROR] [Qnn ExecuTorch]:  <E> Skel failed to process context binary.
[ERROR] [Qnn ExecuTorch]:  <E> Context create from binary failed for deviceId 0 coreId 0 pdId 0 err 5005
[ERROR] [Qnn ExecuTorch]:  <E> Fail to create context from binary with err 5005
[WARNING] [Qnn ExecuTorch]:  <W> sg_stubPtr is not null, skip loadRemoteSymbols
[ERROR] [Qnn ExecuTorch]:  <E> Failed to create context from binary with err 0x138d
[ERROR] [Qnn ExecuTorch]: Can't create context from binary. Error 5005.
```
is it expected?

shewu-quic commented:
> I was able to repro the fp version on my side, but for the 8a8w version, I hit model loading error
> 
> ```
> [ERROR] [Qnn ExecuTorch]:  <E> Skel failed to process context binary.
> [ERROR] [Qnn ExecuTorch]:  <E> Context create from binary failed for deviceId 0 coreId 0 pdId 0 err 5005
> [ERROR] [Qnn ExecuTorch]:  <E> Fail to create context from binary with err 5005
> [WARNING] [Qnn ExecuTorch]:  <W> sg_stubPtr is not null, skip loadRemoteSymbols
> [ERROR] [Qnn ExecuTorch]:  <E> Failed to create context from binary with err 0x138d
> [ERROR] [Qnn ExecuTorch]: Can't create context from binary. Error 5005.
> ```
> 
> is it the same issue you observe from your side?

No. For 8a8w, we could get the compiled graph which is the same as that in fp16. 
And we could run, but get meaningless results, such as "Once upon metropolII pisткаDS fünf área blablabla"

cccclai commented:
turns out I forget the -ptq flag...

what does the performance look like from your side? From the log output, seems like 1-2 toks/s for fp and 0.6 toks/s. Did I miss something?

shewu-quic commented:
> turns out I forget the -ptq flag...I can repro both fp and 8a8w now.
> 
> what does the performance look like from your side? From the log output, seems like 1-2 toks/s for fp and 0.6 toks/s. Did I miss something?

Great! We can start to align each other's results.
Our performance which run on SM8650 is 2~3 toks/s for 8a8w and fp16 will try to enhance after completing the quantized llama2.

cccclai commented:
2~3 toks/s for 8a8w seems still really slow - do we know which part is causing the perf regression? Is delegated part runs reasonably fast and the cpu part is too slow?

shewu-quic committed:
resolve uint16 type and reorder input in runtime

shewu-quic commented:
Hi @cccclai,
We fixed 16a4w accuracy issue which is resolved by [PR 3196](https://github.com/pytorch/executorch/pull/3196).

Results[0]:
Once upon a time, there was a little girl named Lily. She loved to play outside in the sunshine. One day, she saw a big, fluffy cloud in the sky. It looked like a giant cotton candy!
Lily ran inside to tell her mommy about the cloud. "Mommy, mommy, look at the big cloud in the sky! It looks like a giant cotton candy!" she said.

salykovaa commented:
> Hi @cccclai, We fixed 16a4w accuracy issue which is resolved by [PR 3196](https://github.com/pytorch/executorch/pull/3196).
> 
> Results[0]: Once upon a time, there was a little girl named Lily. She loved to play outside in the sunshine. One day, she saw a big, fluffy cloud in the sky. It looked like a giant cotton candy! Lily ran inside to tell her mommy about the cloud. "Mommy, mommy, look at the big cloud in the sky! It looks like a giant cotton candy!" she said.

Dear @shewu-quic @cccclai

Does this resolve [https://github.com/pytorch/executorch/issues/2590](https://github.com/pytorch/executorch/issues/2590)? If so, I will close the issue

salykovaa commented:
Dear @shewu-quic @cccclai,

does [PR 3196](https://github.com/pytorch/executorch/pull/3196) resolve the issue [https://github.com/pytorch/executorch/issues/2590](https://github.com/pytorch/executorch/issues/2590)? If so, I will close the issue. Thank you in advance!

cccclai commented:
> Dear @shewu-quic @cccclai,
> 
> does [PR 3196](https://github.com/pytorch/executorch/pull/3196) resolve the issue #2590? If so, I will close the issue. Thank you in advance!

Thanks for the update and sending the fix! Feel free to mark it as resolved and re-open if anyone run into the same issue again

Commit:
Split llama and modify architecture for performance

Note that this branch is for an example.
llama2 cannot work by this branch.

What we did to optimize performance on HTP is listed:
1. One multihead attentions is transformed to multiple single head.
2. KV-cache is changed to graph I/O. The update is performed in
   qnn_llama_runner.cpp on CPU.
3. llama2 is partitioned to 6 pte files in examples/qualcomm/llama2/composite_llama.py
4. Embedding is quantized. This might need further investigation, e.g.,
   can we move it out of the model on CPU..etc
5. Support u16 and u8 mixed-precision quantization.
6. KV-cache is left as quantized format in graph I/O.
7. RMSNorm is tweaked a bit to reduce the quantization sensitivity.
8. HTP Spill-Fill buffer feature is used among pte files.
9. Convert all Linear layers to Conv2d.
10 Properly set quant_min and quant_max in Observers to offset=128 in
symmetrical quantization.

chiwwang commented:
Rebased as https://github.com/pytorch/executorch/pull/3656/

chiwwang commented:
Please see https://github.com/pytorch/executorch/pull/4142 instead.
